Adding fields

dhruvbhagtani opened the pull request:
Fields1D and Fields2D make field datatypes for variables to represent their location on the grid (centre/face).

dhruvbhagtani commented:
Not really sure if my commenting is correct.

navidcy commented:
OK, let me think a bit on this. But at first glance I'm not sure we need to have `Field1D` and `Field2D`. Perhaps just `Field` and since `Grid` is part of the composite field type the dimensionality can be inferred from whether grid is `::Grid1D` or `::Grid2D`.

navidcy commented on src/fields.jl:
```diff
@@ -0,0 +1,64 @@
+include("grids.jl")
+
+export Grid1D, Grid2D
+
+
```
Delete these lines. No need to include them again. All these are part of the `MixedLayerThermoclineDynamics` module (from the main file in the `src` directory) so the grids have been included and the `Grid1D/2D` have already been exported.

dhruvbhagtani replied:
I see. I'll delete them.

navidcy commented on test/runtests.jl:
```diff
@@ -12,7 +12,7 @@ using Test, MixedLayerThermoclineDynamics
     @test test_dx(grid1D, dx)
     @test test_xF(grid1D, range(0, stop = Lx - dx, length = nx))
-    @test xdomain_length(grid1D, Lx - 0)
+    @test xdomain_length(grid1D, Lx)
```
This is good but it's irrelevant with this PR. It's good to try to have self-contained PRs. Imagine the scenario you want to undo the PR for some reason, then this test for grids will go back to this `Lx-0` weirdness. :)

dhruvbhagtani replied:
I'll keep this in mind from now on.

navidcy reviewed (commented):
A few minor comments. I'd think about this a bit more later and get back to you.

dhruvbhagtani commented:
Yeah, maybe we don't need 1D and 2D, but that's true for Grid also. I don't mind either way, so up to you. I'm more focussed on whether the code is written correctly, so that we can merge it quickly, and I can use it to run some simple cases.

dhruvbhagtani commented:
One more issue: we've defined a number of structures in `fields.jl`, and while defining variables, we'll need to write them as:
`u = Field1D{Face}(udata, grid)`. But Face and Cell need to be defined in the user's file too. We can counter it in two ways:
1. Have a constructor for `Field1D` and `Field2D`, and export that, something we did for `Grids1D` and `Grids2D`.
2. Have `fields.jl` in a module and import the whole thing.

navidcy commented:
Let me think a bit about on this and hopefully we'll finalize today/tomorrow.

navidcy commented:
> One more issue: we've defined a number of structures in `fields.jl`, and while defining variables, we'll need to write:
> `u = Field1D{Face}(udata, grid)`. But `Face` and `Cell` need to be defined in the user's file too. We can counter it in two ways:
> 
> 1. Have a constructor for `Field1D` and `Field2D`, and export that, something we did for `Grids1D` and `Grids2D`.
> 2. Have `fields.jl` in a module and import the whole thing.

--> 1.

I'll make some edits and push.

navidcy commented:
How do you like what I pushed? I'm not 100% that this is better.
We need to add some tests...

navidcy commented:
I think the convention is to define the supertypes in the main package file. That's why I moved `abstract type AbstractLocation end` there.

navidcy commented on src/fields.jl:
```diff
@@ -0,0 +1,33 @@
+"""
+    Field2D(LX, data, grid::Grid1D)
+		
+		Constructs a 2D field of `data` at location `(LX, LY)` on `grid`.
```
Don't use tabs

navidcy commented:
> How do you like what I pushed? I'm not 100% that this is better.
> We need to add some tests...

The truth is that what I did was not working or not functional. I'll push some updates.

navidcy commented:
I actually figured out it was not working when I tested it 🤣

dhruvbhagtani commented:
I changed `fields.jl` because earlier, the user had to explicitly pass `Nothing` as an argument to be able to create a 1D field, whereas now, it's not the case. I think this is what you wanted to implement? Also, I removed `G` as a parameter, because we can explicitly say that the grid needs to be of `Grid1D` or `Grid2D` type, since I created two structures. 

I removed `where` too, because it wasn't required.

navidcy commented:
> I changed `fields.jl` because earlier, the user had to explicitly pass `Nothing` as an argument to be able to create a 1D field, whereas now, it's not the case.

That's not true. I had written the `Field1D` constructor below
```julia
Field1D(LX, data, grid::Grid1D) = Field(LX, Nothing, data, grid)
```
with which you could create a 1D field.

> Also, I removed `G` as a parameter, because we can explicitly say that the grid needs to be of `Grid1D` or `Grid2D` type, since I created two structures.

True, for the reason you mentioning here is not needed. But given a field now, can we understand whether the field is 1D or 2D? I was perhaps thinking ahead and that's why I had included the type of the grid as a field parameter.

Let me have a look the changes you did.

dhruvbhagtani commented:
While coding this up, I also started wondering whether we can change Grids to have only one function available to the user, and based on the number of arguments, the code can itself decide whether it is a `1D` or a `2D` grid. What do you say?

navidcy commented:
Sure, but that's not as urgent at this point. The grids are only constructed once.

Also, there is only one function `Field` now but it has two methods (based on the number of arguments it gets).

navidcy commented:
@dhruvbhagtani have a look what I did and let me know if you have any objections/suggestions.
